<a href="https://colab.research.google.com/github/Nuelky/FE/blob/main/DP_GWP1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Step 1: Put-Call Parity in the Context of the Binomial Tree Model**

1. **Does put-call parity apply for European options? Why or why not?**

Yes. Put-call parity apply for European options.It is based on the principle of no-arbitrage-this ensures that the relationship between the price of a European call option and the price of a European put option(with the same strike price and expiration) will be described in a specific way. Therelationship holds for European options beacuse ir can only be exercised at expiration and this inturns makes it possible to form a perfect synthetic portfolio consisting of a call, a put and the underlying asset or its equivalent eg. a risk-free bond.

We express The put-call parity relationship as-

$$
C + K e^{-rT} = P + S_0
$$

Where,

- $C$ = is the price of the European call option
- $P$ = this is the Price of the European put option
- $S_0$ = and this is the Current price of the underlying asset
- $K$ =K is theStrike price of the option
- $r$ =r represent the Risk-free interest rate
- $T$ = this is the time to maturity

we see (above) the relationship between the prices of the European call and put, the underlying stock and a risk-free bond. There's an absence of arbitrage which ensures that there is no discrepancies in the pricing of these instruments under normal market conditions(Hull, 2015).

The European option is only exercised at maturity and hence the the value of this relationship remains intact up to the expiration date (Black & Scholes, 1973).

2. **Rewrite put-call parity to solve for the call price in terms of everything else.**

Let's rearrange the put-call parity formula which will help us to find the call price ($C$) in terms of the other variables. so,  
$$
C = P + S_0 - K e^{-rT}
$$

This equation shows/means that the call price is determined if we know the put price, the stock price and the strike price(k) discounted to the present value using the risk-free interest rate (Hull, 2015).

3. **Rewrite put-call parity to solve for the put price in terms of everything else.**

Now lets rearrange the equation to find the put price ($P$) in terms of the other variables.

$$
P = C + K e^{-rT} - S_0
$$

we see that if we know the call price we'll be able to calculate the put price using the current stock price, the strike price and the present value of the strike discounted by the risk-free rate (Black & Scholes, 1973).

4. **Does put-call parity apply for American options? Why or why not?**

No , it doesnt.

Putcall parity does not strictly apply to American options and this is for the reason that American options have the feature of early exercise which gives the holder the right to exercise the option at any point before the expiration date.

For American options the ability to exercise early makes the value of a call or put deviate from the strict put-call parity relationship. Put-call parity for American options dno't hold coz the potential benefit of early exercise, moreso forthe  American puts which will become more valuable if the underlying asset's price drops significantly (Cox, Ross, & Rubinstein, 1979).


**Lets Pricing European Calls and Puts Using a Binomial Tree**

# We have-
- $S_0 = 100$: Initial stock price
- $r = 5\%$: Risk-free interest rate
- $\sigma = 20\%$: Volatility
- $T = 3$ months (0.25 years)

The options to be priced are At-the-Money (ATM).

**5. Price an ATM European Call and Put Using a Binomial Tree**


In [ ]:
import numpy as np

# Our parameters
S0 = 100  # This is Initial stock price
K = 100   # Strike price (ATM)
r = 0.05  # Risk-free interest rate
T = 0.25  # Time to maturity (3 months)
sigma = 0.2  # Volatility (20%)
N = 100  # Number of time steps
dt = T / N  # Time increment
u = np.exp(sigma * np.sqrt(dt))  # Our Up factor
d = 1 / u  # Down factor
p = (np.exp(r * dt) - d) / (u - d)  # Risk-neutral probability

# Initialize arrays for option values
call_values = np.zeros(N+1)
put_values = np.zeros(N+1)

# Stock prices at maturity
stock_prices = np.array([S0 * u**j * d**(N-j) for j in range(N+1)])

# Calculate option payoffs at maturity
call_values = np.maximum(stock_prices - K, 0)
put_values = np.maximum(K - stock_prices, 0)

# Backward induction to calculate the option prices at time 0
for i in range(N-1, -1, -1):
    call_values = np.exp(-r * dt) * (p * call_values[1:] + (1 - p) * call_values[:-1])
    put_values = np.exp(-r * dt) * (p * put_values[1:] + (1 - p) * put_values[:-1])

# European call and put prices at time 0
european_call_price = call_values[0]
european_put_price = put_values[0]

print(f"European Call Price: {european_call_price:.2f}")
print(f"European Put Price: {european_put_price:.2f}")


European Call Price: 4.61
European Put Price: 3.36



**Question 5a. Choose the Number of Steps in the Tree to Achieve Reliable Estimates**

Let's use 100 steps in the binomial tree model.

We know that he number of steps in a binomial tree affects the accuracy of the option price estimates.the model approximates better the continuous movement of the stock with more steps while on the other hand there is a balance between computational complexity and accuracy.

**Why 100 Steps?**

 Cox, Ross & Rubinstein (1979) indicated that 100 steps provide a good approximation to the true price more so for the options with relatively short maturities (in our case we have 3 months) while f ewer steps will or may lead to less accurate estimates for the options close to expiration while also on the other hand too many steps will increase computational time without significantly improving the result.

**b. Overall Process of the Binomial Tree Model**

What is binomial tree model and how does it work?-It is a widely used in pricing options where it divides the time to maturity into discrete intervals and models the underlying asset price as moving either up or down at each step- this forms a tree-like structure of potential outcomes.

The general process is->

1. **We start by Defining our  Parameters**

We begin by defining the up ($u$) and down ($d$) factors based on the volatility ($\sigma$) and the time step ($dt$) which will help us determine how much the stock price can move up or down at each step...

$$
u = e^{\sigma \sqrt{dt}} \quad \text{and} \quad d = \frac{1}{u}
$$

where $dt = \frac{T}{N}$ is the length of each time step.

2. **Next we Calculate the Risk-Neutral Probability**

We us the risk-neutral probability ($p$) to discount future option payoffs back to the present

$$
p = \frac{e^{r \cdot dt} - d}{u - d}
$$

This probability will ensures that the expected price at each node is consistent with the no-arbitrage condition in option pricing.

3. **Then the next step will be to Generate Stock Prices at Maturity**

-We now calculate the possible stock prices at each node using the up and down factors at the final step of the tree (maturity),

4. **Now we Calculate the Option Payoffs at Maturity**

-  the payoff at each node for the call option is-

$$
\max(S_T - K, 0)
$$

 while the payoff for the put option is

$$
\max(K - S_T, 0)
$$

5. **The next step is known as Backward Induction**


Here we discount the option payoffs step-by-step back to the present using the risk-neutral probabilities and the risk-free rate.

We ustThe formula -

$$
V = e^{-r \cdot dt} (p \cdot V_{\text{up}} + (1 - p) \cdot V_{\text{down}})
$$

we repeat this process until we reach time 0 where we calculate the initial option price.

**Our Results from the Binomial Tree**

We obtained the  following prices for the ATM European call and put options using the binomial tree model with 100 steps(see the code above)

- **European Call Price:** 4.61
- **European Put Price:** 3.36

There is a slight difference between the call and put prices as we expected due to the put-call parity relationship and the nature of ATM options where the stock price is equal to the strike price.

Our results above are consistent with theoretical expectations for an At-the-Money (ATM) option where- the time value and volatility contribute significantly to the option's premium (Hull, 2015).


##Question 6: Compute the Greek Delta for the European Call and European Put at Time 0

In [ ]:
# Parameters for small stock price changes
delta_S = 1  # Small change in stock price
S_up = S0 + delta_S  # Stock price moves up
S_down = S0 - delta_S  # Stock price moves down

# Recalculating stock prices at maturity for S_up and S_down
stock_prices_up = np.array([S_up * u**j * d**(N-j) for j in range(N+1)])
stock_prices_down = np.array([S_down * u**j * d**(N-j) for j in range(N+1)])

# Recalculate option payoffs at maturity
call_values_up = np.maximum(stock_prices_up - K, 0)
call_values_down = np.maximum(stock_prices_down - K, 0)
put_values_up = np.maximum(K - stock_prices_up, 0)
put_values_down = np.maximum(K - stock_prices_down, 0)

# Backward induction for call and put option prices with S_u and S_down
for i in range(N-1, -1, -1):
    call_values_up = np.exp(-r * dt) * (p * call_values_up[1:] + (1 - p) * call_values_up[:-1])
    call_values_down = np.exp(-r * dt) * (p * call_values_down[1:] + (1 - p) * call_values_down[:-1])
    put_values_up = np.exp(-r * dt) * (p * put_values_up[1:] + (1 - p) * put_values_up[:-1])
    put_values_down = np.exp(-r * dt) * (p * put_values_down[1:] + (1 - p) * put_values_down[:-1])

# Delta calculation for call and put
delta_call = (call_values_up[0] - call_values_down[0]) / (2 * delta_S)
delta_put = (put_values_up[0] - put_values_down[0]) / (2 * delta_S)

print(f"Delta for European Call: {delta_call:.4f}")
print(f"Delta for European Put: {delta_put:.4f}")


Delta for European Call: 0.5693
Delta for European Put: -0.4307


**WHat is Delta?**
this is a measure of how the price of an option changes in response to small changes in the underlying asset's price. It (Delta) represents the sensitivity of the option’s price with respect to movements in the underlying stock price.

We will calculate the Delta for both the European call and put options using a binomial tree approach.

**Queston 6a. Delta Comparison for the European Call and Put Options**

We are using the binomial tree model and adjusting the underlying stock price by a small increment ($\Delta S = 1$) where we computed the following Delta values for the European call and put options(See the code above)-

- **Delta for European Call:** 0.5693
- **Delta for European Put:** -0.4307

These values are indicating/showing how much the price of the call and put options will change for a small change in the stock price.

**b. Analysis of Delta for European Call and Put Options**

We have a positive Delta for the European Call (0.5693) which tells us  that- as the underlying stock price increases by 1 unit the call option's price increases by approximately 0.5693 units. For sure this is consistent with the behavior of call options where we know an an increase in the stock price will lead to an increase in the value of the option.

The Delta is slightly higher due to facts like teh market conditions (Hull, 2015). However, for ATM (At-the-Money) call optionsDelta is normally close to 0.5, but in this case we have0.5693,

We also have a negative Delta for the European Put(-0.4307) which here signifies that- for every 1 unit increase in the stock price, the price of the put option decreases by 0.4307 units. Here also, this negative value aligns with the behavior of put options where we expect that the option’s value decreases as the underlying stock price rises. The Delta for an ATM put is generally/normally near -0.5- However in our case the Delta is -0.43 which means that there's a slight imbalance in sensitivity to stock price changes.

- The magnitude of the Delta values is influenced by the option’s time to maturity and the volatility of the underlying asset (Hull, 2015). Call options will exhibit positive Delta values while on the other hand  put options will exhibit negative Delta values.Options are sensitive to small price changes in the underlying stock which is expected for ATM options and as show by the magnititude of values above.


##Question 7: Sensitivity to Volatility (Vega)

**Q.7: Sensitivity to Volatility (Vega)**

**What is Vega?**

- This is the sensitivity of an option's price to changes in the volatility of the underlying asset.

We will calculate how the prices of the European call and put options change when the volatility increases by 5%, that is from 20% to 25%.

**Q 7a. Compute the Sensitivity of the Call and Put Option Prices to a 5% Increase in Volatility**

Here-We will recalculate the option prices using a volatility of 25%-- then we will compare them with the original prices calculated at 20% volatility.

We use the formula below to compute Vega-

$$
Vega = \frac{V(\sigma_{\text{new}}) - V(\sigma_{\text{old}})}{\sigma_{\text{new}} - \sigma_{\text{old}}}
$$

Where:

- $V(\sigma_{\text{new}})$ is the option price with the new volatility ($\sigma_{\text{new}} = 25\%$).
- $V(\sigma_{\text{old}})$ is the option price with the old volatility ($\sigma_{\text{old}} = 20\%$).

Let's usse this formula in python and calculate the sensitivity (Vega) of the option prices to changes in volatility.


In [ ]:
# We Increase volatility to 25%
sigma_new = 0.25  # New volatility (25%)
u_new = np.exp(sigma_new * np.sqrt(dt))  # New up factor
d_new = 1 / u_new  # New down factor
p_new = (np.exp(r * dt) - d_new) / (u_new - d_new)  # New risk-neutral probability

# Lets Reinitialize arrays for new volatility
call_values_new = np.zeros(N+1)
put_values_new = np.zeros(N+1)

# Now we Recalculate stock prices at maturity for new volatility
stock_prices_new = np.array([S0 * u_new**j * d_new**(N-j) for j in range(N+1)])

# Here we Calculate option payoffs at maturity for new volatility
call_values_new = np.maximum(stock_prices_new - K, 0)
put_values_new = np.maximum(K - stock_prices_new, 0)

# Now-Backward induction to calculate new option prices at time 0 with increased volatility
for i in range(N-1, -1, -1):
    call_values_new = np.exp(-r * dt) * (p_new * call_values_new[1:] + (1 - p_new) * call_values_new[:-1])
    put_values_new = np.exp(-r * dt) * (p_new * put_values_new[1:] + (1 - p_new) * put_values_new[:-1])

# New option prices with 25% volatility
new_call_price = call_values_new[0]
new_put_price = put_values_new[0]

# And our Vega calculation
vega_call = (new_call_price - european_call_price) / (sigma_new - sigma)
vega_put = (new_put_price - european_put_price) / (sigma_new - sigma)

print(f"Vega for European Call: {vega_call:.4f}")
print(f"Vega for European Put: {vega_put:.4f}")


Vega for European Call: 19.6189
Vega for European Put: 19.6189


As we defined above, Vega measures the sensitivity of an option's price to changes in the volatility of the underlying asset.

**a. Sensitivity of Call and Put Option Prices to a 5% Increase in Volatility**

We've recalculated the option prices with the increased volatility of 25% Using a binomial tree model,. The resulting Vega values represent the change in option prices for each 1% increase in volatility as the output above we see-

- **Vega for European Call:** 19.6189
- **Vega for European Put:** 19.6189

The values shows - the call and put option prices (both) are highly sensitive to changes in volatility shown by an increase of 19.62 units for every 1% increase in volatility.

**b. Comment on the Differential Impact of This Change for Call and Put Options**

- **For European Call Option:**

The Vega of the call option shows/indicates that a 5% increase in volatility will lead to a (substantial) increase in the call option’s price. This means higher volatility increases the likelihood of upward movements in the stock price which will in turn benefit the call option holders.

- **European Put Option:**
Here, the Vega for the put option is  as for the call option which si to  mean the put option's price will also increase by 19.62 units for each 1% increase in volatility.Therefore this is because higher volatility raises the chance of the stock price dropping below the strike price and which also benefits the put option holder.

Both the call and put options exhibit the same Vega values because they are ATM (At-the-Money) options where the sensitivity to volatility is at its peak.

Our result is consistent with the behavior of options-ATM options are more sensitive to changes in volatility compared to in-the-money (ITM) or the out-of-the-money (OTM) options (Hull, 2015).


 ## Team Member B on American Options

**Question 8.To price an ATM American call and put option using a binomial tree:**

a. Here we chose n=100 steps. Given that as the number of steps increases, the binomial model price converges to the Black-Scholes model price. And to ease computation, 100 steps is ideal.

b. the overall process:

- First setting up the parameters.
- Followed by calculating the up (u) and down (d) multipliers using the given volatility and time per step.
- The probability q for the up is move is calculated using u, d, and the risk-free rate.
- A binomial tree of stock prices is built. At the end of the tree (maturity), the option payoffs are calculated for both call and put options.

Finally, we recursively calculate the option prices backwards through the tree, considering the early exercise for American options.

In [ ]:
import numpy as np
import math

# Our parameters
S_0 = 100
r = 0.05
σ = 0.20
T = 0.25
steps = 100
Δt = T/steps
u = math.exp(σ * math.sqrt(Δt))
d = 1/u
discount = math.exp(-r * Δt)

# To Generate binomial stock price tree
stock_tree = np.zeros((steps + 1, steps + 1))
for j in range(steps+1):
    for i in range(j+1):
        stock_tree[i, j] = S_0 * (u ** i) * (d ** (j - i))

# Initialize option values tree for both call and put
call_tree = np.zeros((steps + 1, steps + 1))
put_tree = np.zeros((steps + 1, steps + 1))

# Compute option values at terminal nodes
for i in range(steps + 1):
    call_tree[i, steps] = max(0, stock_tree[i, steps] - S_0)
    put_tree[i, steps] = max(0, S_0 - stock_tree[i, steps])

for j in range(steps-1, -1, -1):
    for i in range(j+1):
        call_tree[i, j] = max(stock_tree[i, j] - S_0,
                              discount * (0.5 * call_tree[i, j+1] + 0.5 * call_tree[i+1, j+1]))
        put_tree[i, j] = max(S_0 - stock_tree[i, j],
                             discount * (0.5 * put_tree[i, j+1] + 0.5 * put_tree[i+1, j+1]))

print(f"American ATM Call Option Price: {call_tree[0, 0]:.2f}")
print(f"American ATM Put Option Price: {put_tree[0, 0]:.2f}")

American ATM Call Option Price: 4.19
American ATM Put Option Price: 3.74


**9.Computing the greeks**

Here we replicated what we did previously for the European option.

Using the previous binomial tree model, we can compute Δ at time 0 using the option prices at the first up-move and the initial option prices.

In [ ]:
# Calculate Deltas
delta_call = (call_tree[1, 1] - call_tree[0, 1]) / (stock_tree[1, 1] - stock_tree[0, 1])
delta_put = (put_tree[1, 1] - put_tree[0, 1]) / (stock_tree[1, 1] - stock_tree[0, 1])

print(f"Delta for the American ATM Call Option: {delta_call:.4f}")
print(f"Delta for the American ATM Put Option: {delta_put:.4f}")

Delta for the American ATM Call Option: 0.5357
Delta for the American ATM Put Option: -0.4644


Comments:

*How do they compare?*
- Delta for the call option is positive and less than 1 where as the Delta for the put option is negative and greater than -1.

*Comment briefly on the differences and signs of Delta for both options. Delta for Call Option:*
- The Delta of a call option will range from 0 to 1. This implies that for every unit increase in the stock price, the call option price will increase by its Delta. Given that the American call option gives the holder the right but not the obligation to buy the underlying stock, as the stock price increases, the call option value increases.
- Delta for Put Option: The Delta of a put option will always lie from -1 and 0. This shows that for every unit increase in the stock price, the put option price will decrease by its absolute Delta value. We know that the American put option gives the holder the right to sell the stock, as the stock price rises, the put option becomes less valuable as one could sell the stock at a higher market price.
- What does delta proxy for? Delta shows how sensitive the option's price is to a change in the price of the underlying asset. It's basically a measure of the option's exposure to the movement in the underlying asset.

*Why does it make sense to obtain a positive/negative delta for each option?*

- For a call option, a positive delta makes sense because as the stock price increases, the value of the call option i.e. the right to buy equally increases. Hence one would benefit from an increase in the stock's price. Also, For a put option, a negative delta is logical, this is because as the stock price increases, the value of the put option which is right to sell decreases. If the stock's price is high, the value of having an option to sell it at a lower price goes down. This relationship between option values and stock price movements is very important in understanding the hedging and risk management of options.

**Question 10: In computing Option sensitivity (Vega)**

- To calculate the sensitivity of the option prices to a change in volatility, It is important to know that just as we did for the European options, we applied the same to the American option. so, price the options using the original volatility (σ=20%) and price the options again using a 5% increase in volatility, hence (σ=25%).
- Lastly we compute the difference in the option prices from the two volatilities to determine the sensitivity (vega).

In [ ]:
from math import exp, sqrt

σ_new = 0.25
u_new = math.exp(σ_new * math.sqrt(Δt))
d_new = 1/u_new


stock_tree_new = np.zeros((steps + 1, steps + 1))
for j in range(steps+1):
    for i in range(j+1):
        stock_tree_new[i, j] = S_0 * (u_new ** i) * (d_new ** (j - i))


call_tree_new = np.zeros((steps + 1, steps + 1))
put_tree_new = np.zeros((steps + 1, steps + 1))
for i in range(steps + 1):
    call_tree_new[i, steps] = max(0, stock_tree_new[i, steps] - S_0)
    put_tree_new[i, steps] = max(0, S_0 - stock_tree_new[i, steps])

for j in range(steps-1, -1, -1):
    for i in range(j+1):
        call_tree_new[i, j] = max(stock_tree_new[i, j] - S_0,
                                  discount * (0.5 * call_tree_new[i, j+1] + 0.5 * call_tree_new[i+1, j+1]))
        put_tree_new[i, j] = max(S_0 - stock_tree_new[i, j],
                                 discount * (0.5 * put_tree_new[i, j+1] + 0.5 * put_tree_new[i+1, j+1]))


vega_call = (call_tree_new[0, 0] - call_tree[0, 0]) / 0.05
vega_put = (put_tree_new[0, 0] - put_tree[0, 0]) / 0.05

print(f"Sensitivity of American ATM Call Option Price to 5% increase in volatility: {vega_call:.2f}")
print(f"Sensitivity of American ATM Put Option Price to 5% increase in volatility: {vega_put:.2f}")

Sensitivity of American ATM Call Option Price to 5% increase in volatility: 22.70
Sensitivity of American ATM Put Option Price to 5% increase in volatility: 17.57


*How do prices change with respect to the change in volatility?*
- The computed values tells us by how much the option prices change for a 5% increase in volatility.

*Comment on the potential differential impact of this change for call and put options.*
- Both call and put option prices increase with volatility. The increased volatility implies that there is high potential chances of price swings in the underlying asset. This makes both the call and put options more valuable due to the high chance of them ending up in-the-money. Vega is positive for both calls and puts, meaning that both option types benefit from an increase in volatility. This is because the more volatile the stock, the higher the probability that it could make a large move either upward or downward. Although both option types benefit from increased volatility, the exact magnitude of the increase might be different for calls and puts. This difference can be affected by factors such as the moneyness of the options, the time to expiration, and interest rates.

## Step 2 – Unless stated otherwise, keep the same input parameters as for Step 1.
**Team Member B will now work with European options using a trinomial tree:**

**Question 15**.Price European options using a trinomial tree, the idea is to Select 5 strike prices corresponding to different levels of moneyness (thats Deep OTM, OTM, ATM, ITM, and Deep ITM).

second, We will Use a trinomial tree to price the American-style call options with these 5 strikes.

Then we will Comment on the trend in option prices as moneyness changes (that is to mean as the strike price moves further from the current stock price).

In [ ]:
import numpy as np

# Parameters
S0 = 100  # Current stock price
r = 0.05  # Risk-free interest rate
T = 0.25  # Time to maturity (3 months)
sigma = 0.2  # Volatility
N = 100  # Number of steps in the tree

# Strike prices for different moneyness levels
strike_prices = [110, 105, 100, 95, 90]  # Deep OTM, OTM, ATM, ITM, Deep ITM

# Calculations for trinomial tree parameters
dt = T / N  # Time increment
u = np.exp(sigma * np.sqrt(2 * dt))  # Up factor
d = 1 / u  # Down factor
p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2  # Probability of up movement
p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2  # Probability of down movement
p_m = 1 - p_u - p_d  # Probability of no movement

def trinomial_tree_european_call(S0, K, r, sigma, T, N):
    dt = T / N
    u = np.exp(sigma * np.sqrt(2 * dt))
    d = 1 / u
    p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2
    p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2
    p_m = 1 - p_u - p_d

    # Initialize asset prices at maturity
    stock_prices = np.zeros((N + 1, 2 * N + 1))
    stock_prices[0, N] = S0

    for i in range(1, N + 1):
        for j in range(N - i, N + i + 1, 2):
            stock_prices[i, j] = stock_prices[i - 1, j - 1] * u
            stock_prices[i, j - 2] = stock_prices[i - 1, j - 1] * d

    # Initialize option values at maturity (European payoff)
    call_values = np.maximum(stock_prices[N] - K, 0)

    # Work backwards through the tree (no early exercise)
    for i in range(N - 1, -1, -1):
        for j in range(N - i, N + i + 1, 2):
            call_values[j] = np.exp(-r * dt) * (
                p_u * call_values[j + 1] + p_m * call_values[j] + p_d * call_values[j - 1]
            )

    return call_values[N]

# Pricing European call options for the selected strikes
for K in strike_prices:
    call_price = trinomial_tree_european_call(S0, K, r, sigma, T, N)
    print(f"European Call->> where K={K} is = {call_price:.2f}")


European Call->> where K=110 is = 0.35
European Call->> where K=105 is = 0.96
European Call->> where K=100 is = 2.17
European Call->> where K=95 is = 4.20
European Call->> where K=90 is = 6.91


Using the Trinomial Tree to Price the European Call Option for 5 Strikes

Here, we selected five strike prices to represent different levels of moneyness for European call options. The parameters are as follows:

- 𝑆0 =100: Current stock price
- r=5%: Risk-free interest rate
- σ=20%: Volatility
- T=0.25 years (Time to maturity: 3 months)

The strike prices are:
- Deep Out-of-the-Money (OTM): K = 110
- Out-of-the-Money (OTM): K = 105
- At-the-Money (ATM): K = 100
- In-the-Money (ITM): K = 95
- Deep In-the-Money (ITM): K = 90

Applying the trinomial tree model to these strikes for European call options, we obtained the following prices:

- Deep OTM (K = 110): 0.35
- OTM (K = 105): 0.96
- ATM (K = 100): 2.17
- ITM (K = 95): 4.20
- Deep ITM (K = 90): 6.91

*Comment on the Trend in Option Prices as Moneyness Changes*

As expected, the value of the European call option increases as the strike price decreases relative to the current stock price (i.e., as the option moves from out-of-the-money to in-the-money).

This trend is intuitive because a lower strike price increases the probability of profitable exercise at maturity. Therefore, options that are deep in-the-money have the highest value since they are more likely to finish in-the-money. However, unlike American options, European options cannot be exercised early, which limits their value increase compared to American options at similar levels of moneyness.

**Question 16. Price European put Option Using a Trinomial Tree**

In [ ]:
def trinomial_tree_european_put(S0, K, r, sigma, T, N):
    dt = T / N
    u = np.exp(sigma * np.sqrt(2 * dt))
    d = 1 / u
    p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2
    p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2
    p_m = 1 - p_u - p_d

    # Initialize asset prices at maturity
    stock_prices = np.zeros((N + 1, 2 * N + 1))
    stock_prices[0, N] = S0

    for i in range(1, N + 1):
        for j in range(N - i, N + i + 1, 2):
            stock_prices[i, j] = stock_prices[i - 1, j - 1] * u
            stock_prices[i, j - 2] = stock_prices[i - 1, j - 1] * d

    # Initialize option values at maturity (European payoff for puts)
    put_values = np.maximum(K - stock_prices[N], 0)

    # Work backwards through the tree (no early exercise)
    for i in range(N - 1, -1, -1):
        for j in range(N - i, N + i + 1, 2):
            put_values[j] = np.exp(-r * dt) * (
                p_u * put_values[j + 1] + p_m * put_values[j] + p_d * put_values[j - 1]
            )

    return put_values[N]

# Pricing European put options for the selected strikes
put_strike_prices = [90, 95, 100, 105, 110]  # Deep OTM, OTM, ATM, ITM, Deep ITM
for K in put_strike_prices:
    put_price = trinomial_tree_european_put(S0, K, r, sigma, T, N)
    print(f"European Put->> where K={K} is = {put_price:.2f}")


European Put->> where K=90 is = 29.99
European Put->> where K=95 is = 32.24
European Put->> where K=100 is = 35.16
European Put->> where K=105 is = 38.92
European Put->> where K=110 is = 43.27


Here, we selected five strike prices to represent different levels of moneyness for European put options. The parameters are as follows:

- 𝑆0 = 100 : Current stock price
- 𝑟 = 5% Risk-free interest rate
- σ = 20%: Volatility
- T = 0.25 years (Time to maturity: 3 months)

The strike prices are:

- Deep Out-of-the-Money (OTM):K = 90
- Out-of-the-Money (OTM): K = 95
- At-the-Money (ATM): K = 100
- In-the-Money (ITM): K = 105
- Deep In-the-Money (ITM): K = 110

Applying the trinomial tree model to these strikes for European put options, we obtained the following prices:

- Deep OTM (K = 90): 29.99
- OTM (K = 95): 32.24
- ATM (K = 100): 35.16
- ITM (K = 105): 38.92
- Deep ITM (K = 110): 43.27

*Comment on the Trend in Option Prices as Moneyness Changes*

The value of the European put option increases as the strike price rises relative to the current stock price (i.e., as the option moves from out-of-the-money to in-the-money).

This trend is logical because a higher strike price increases the probability that the option will finish in-the-money, resulting in a profitable exercise at maturity. Consequently, options that are deep in-the-money have the highest values, as they are most likely to be exercised profitably. The payoff potential for European put options aligns with the intrinsic value increase as the strike price moves further above the stock price.

Question 17: Pricing American Call Options Using a Trinomial Tree

**Our Objectives here are:**

- T0 Select 5 strike prices corresponding to different levels of moneyness (thats Deep OTM, OTM, ATM, ITM, and Deep ITM).

- We will Use a trinomial tree to price the American-style call options with these 5 strikes.
- Then we will Comment on the trend in option prices as moneyness changes (that is to mean as the strike price moves further from the current stock price).

In [ ]:
import numpy as np

# Parameters
S0 = 100  # Current stock price
r = 0.05  # Risk-free interest rate
T = 0.25  # Time to maturity (3 months)
sigma = 0.2  # Volatility
N = 100  # Number of steps in the tree

# Strike prices for different moneyness levels
strike_prices = [110, 105, 100, 95, 90]  # Deep OTM, OTM, ATM, ITM, Deep ITM

dt = T / N  # Time increment
u = np.exp(sigma * np.sqrt(2 * dt))  # Up factor
d = 1 / u  # Down factor
p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2  #The Probability of up movement
p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2  #The Probability of down movement
p_m = 1 - p_u - p_d  #The Probability of no movement

def trinomial_tree_american_call(S0, K, r, sigma, T, N):
    dt = T / N
    u = np.exp(sigma * np.sqrt(2 * dt))
    d = 1 / u
    p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2
    p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2
    p_m = 1 - p_u - p_d

    # Initialize asset prices at maturity
    stock_prices = np.zeros((N + 1, 2 * N + 1))
    stock_prices[0, N] = S0

    for i in range(1, N + 1):
        for j in range(N - i, N + i + 1, 2):
            stock_prices[i, j] = stock_prices[i - 1, j - 1] * u
            stock_prices[i, j - 2] = stock_prices[i - 1, j - 1] * d

    # Initializing option values at maturity
    call_values = np.maximum(stock_prices[N] - K, 0)

    # Work backwards through the tree
    for i in range(N - 1, -1, -1):
        for j in range(N - i, N + i + 1, 2):
            hold_value = np.exp(-r * dt) * (
                p_u * call_values[j + 1] + p_m * call_values[j] + p_d * call_values[j - 1]
            )
            call_values[j] = np.maximum(hold_value, stock_prices[i, j] - K)  # Early exercise

    return call_values[N]

# Pricing American call options for the selected strikes
for K in strike_prices:
    call_price = trinomial_tree_american_call(S0, K, r, sigma, T, N)
    print(f"American Call->> where K={K} is = {call_price:.2f}")


American Call->> where K=110 is = 0.52
American Call->> where K=105 is = 1.44
American Call->> where K=100 is = 3.26
American Call->> where K=95 is = 6.29
American Call->> where K=90 is = 10.36


**a. Using the Trinomial Tree to Price the American Call Option for 5 Strikes**

Here we have selected five strike prices to represent different levels of moneyness for the American call options. The parameters are as follows as we know:

- $S_0 = 100$: Current stock price
- $r = 5\%$: Risk-free interest rate
- $\sigma = 20\%$: Volatility
- $T = 3$ months (Time to maturity = 0.25 years)

The strike prices are-:

- **Deep OTM**: $K = 110$
- **OTM**: $K = 105$
- **ATM**: $K = 100$
- **ITM**: $K = 95$
- **Deep ITM**: $K = 90$

We"ll use a **trinomial tree** to price the American call options at these five strikes. The trinomial tree model will account for three possible price movements at each step (up, down, or no movement) and will includes the possibility of **early exercise** (we know thatit is a feature of American options).


#### After applying the trinomial tree model we obtained the following prices for the American call options-

- **Deep OTM (K = 110):** 0.52
- **OTM (K = 105):** 1.44
- **ATM (K = 100):** 3.26
- **ITM (K = 95):** 6.29
- **Deep ITM (K = 90):** 10.36

**17b. Comment on the Trend in Option Prices as Moneyness Changes**

 As we expected the value of the American call option will increases as the strike price decreases relative to the stock price (this is to say-as the option moves from **OTM** to **ITM**).

 This trend is intuitive;the lower the strike price the higher the probability that the option will be exercised profitably ( to mean tthat it will finish **in-the-money**).

 The deepest **in-the-money** options have the highest value-they are most likely to be exercised early; as as American-style options (Hull, 2015).


#Q 18: Pricing American Put Options Using a Trinomial Tree

**For Question 18, our Objectives are->**

- We will need to Select 5 strike prices corresponding to different levels of moneyness (thats is Deep OTM, OTM, ATM, ITM, Deep ITM).
- We'll Use a trinomial tree to price American-style put options for these strike prices.
-Then we"ll Comment on the trend in option prices as moneyness changes (Thats as the strike price moves further from the current stock price).

In [ ]:
def trinomial_tree_american_put(S0, K, r, sigma, T, N):
    dt = T / N
    u = np.exp(sigma * np.sqrt(2 * dt))
    d = 1 / u
    p_u = ((np.exp(r * dt / 2) - d) / (u - d)) ** 2
    p_d = ((u - np.exp(r * dt / 2)) / (u - d)) ** 2
    p_m = 1 - p_u - p_d

    # Initialize asset prices at maturity
    stock_prices = np.zeros((N + 1, 2 * N + 1))
    stock_prices[0, N] = S0

    for i in range(1, N + 1):
        for j in range(N - i, N + i + 1, 2):
            stock_prices[i, j] = stock_prices[i - 1, j - 1] * u
            stock_prices[i, j - 2] = stock_prices[i - 1, j - 1] * d

    # Initialize option values at maturity
    put_values = np.maximum(K - stock_prices[N], 0)

    # Work backwards through the tree
    for i in range(N - 1, -1, -1):
        for j in range(N - i, N + i + 1, 2):
            hold_value = np.exp(-r * dt) * (
                p_u * put_values[j + 1] + p_m * put_values[j] + p_d * put_values[j - 1]
            )
            put_values[j] = np.maximum(hold_value, K - stock_prices[i, j])  # Early exercise

    return put_values[N]

# Pricing American put options for the selected strikes
for K in strike_prices:
    put_price = trinomial_tree_american_put(S0, K, r, sigma, T, N)
    print(f"American Put Price (K={K})is equals to {put_price:.2f}")


American Put Price (K=110)is equals to 43.27
American Put Price (K=105)is equals to 38.92
American Put Price (K=100)is equals to 35.16
American Put Price (K=95)is equals to 32.24
American Put Price (K=90)is equals to 29.99


**18.a. Using the Trinomial Tree to Price the American Put Option for 5 Strikes**

We selecte 5 strike prices to represent different levels of moneyness for the American put options. The parameters are--

- $S_0 = 100$: Current stock price
- $r = 5\%$: Risk-free interest rate
- $\sigma = 20\%$: Volatility
- $T = 3$ months (Time to maturity = 0.25 years)

The strike prices are-

- **Deep OTM**: $K = 110$
- **OTM**: $K = 105$
- **ATM**: $K = 100$
- **ITM**: $K = 95$
- **Deep ITM**: $K = 90$

We''ll use a **trinomial tree** to price the American put options at these five strikes. The trinomial tree model will accounts for three possible price movements at each step (again that will be up, down, or no movement) and also it will includes the possibility of **early exercise** (as in American options).



We apply the trinomial tree model to get the pricesfor the American put options as below

- **Deep OTM (K = 110):** 43.27
- **OTM (K = 105):** 38.92
- **ATM (K = 100):** 35.16
- **ITM (K = 95):** 32.24
- **Deep ITM (K = 90):** 29.99

**b. Comment on the Trend in Option Prices as Moneyness Changes**

Again as we expected the value of the American put  will increases as the strike price increase relative to the stock price (that is as the option moves from **ITM** to **OTM**).

The trend makes sense as- the higher the strike price the higher the probability that the option will finish **in-the-money** and also be exercised early which in turn increases its value.

- **Deep OTM (K = 110)**- has the highest value (43.27); the strike price is significantly higher than the stock price. This means that it is very likely the option will be exercised early.
- **OTM (K = 105)** -thuis shows a decrease in value (38.92) as the strike price moves closer to the stock price which reduces the probability of early exercise.
- **ATM (K = 100)** this has a balanced price (35.16);the strike price is equal to the stock price. This is a  reflection of the option's probability in finishing either in-the-money or out-of-the-money.
- **ITM (K = 95)** and **Deep ITM (K = 90)** have lower values(that"s 32.24 and 29.99 respectively). This is because the likelihood of these options being exercised early will decreases as the strike price moves further below the stock price.

The trend shows and confirms to us that- as the strike price increases the value of the American put option will also increase and its because the probability of profitable early exercise also increases and as we know is a key feature of American-style options (Hull, 2015).


##Step 3

**Question 25.Dynamic Delta Hedging.**

S_0=180, r =2%, sigma=25%, T=6
months, K = 182:

###A. European put option using a 3-step binomial tree

**Already we have been given the Data blw**

- $S_0 = 180$:This is Current stock price
- $K = 182$: K is the Strike price
- $r = 2\% = 0.02$: ris the Risk-free interest rate
- $\sigma = 25\% = 0.25$: Volatility
- $T = 6$ months = 0.5 years
- We have 3 steps in the binomial tree

**Let's start with Calculating Time Step ($\Delta t$)**

The time step for each period is given by--

$$
\Delta t = \frac{T}{N} = \frac{0.5}{3} = 0.167 \text{ years}
$$

**Then Lets Calculate Up and Down Factors ($u$ and $d$)**

The factors are based on the volatility ($\sigma$) and the time step ($\Delta t$) and we know that-


$$
u = e^{\sigma \sqrt{\Delta t}} = e^{0.25 \times \sqrt{0.1667}} = e^{0.10206} = 1.1075
$$

 we know that $d=\frac{1}{u}$, so let"s use it to get u:

$$
d = \frac{1}{u} = \frac{1}{1.1075} = 0.9029
$$

**Now we The Calculate Risk-Neutral Probability ($p$)**

The risk-neutral probability -we'll calculate it using the risk-free interest rate ($r$) and the up and down (that is ($u$) and ($d$))factors:

$$
p = \frac{e^{r \Delta t} - d}{u - d}
$$

First, lets calculate $e^{r \Delta t}$--

$$
e^{r \Delta t} = e^{0.02 \times 0.1667} = e^{0.003334} = 1.0033
$$

Now we calculate $p$-

$$
   $$Using the values,$$   p = \frac{1.0033 - 0.9029}{1.1075 - 0.9029} = \frac{0.1004}{0.2046} =0.4908
$$

Thus, **$p = 0.4908$.**


**Our Next Step is Stock Prices at Each Node**

We now calculate the possible stock prices at each node in the binomial tree, We starting from $S_0 = 180$.

- **Step 0 ($Initial$):** $S_0 = 180$

- **Step 1 ( $u $  and $d$):**
  - $S_{\text{up}} = 180 \times 1.1075 = 199.35$
  - $S_{\text{down}} = 180 \times 0.9029 = 162.53$
- **Step 2 (Up and Down):**
  - $S_{\text{up-up}} = 199.35 \times 1.1075 = 220.87$
  - $S_{\text{down-down}} = 162.53 \times 0.9029 = 146.70$
  - $S_{\text{up-down}} = S_{\text{down-up}} = 180$

**Next Step: We need to know Payoffs at Maturity**

At maturity- we'll calculate the payoff of the European put option as $\max(K - S, 0)$.

- For $S=220.87$: Payoff = $\max(182 - 220.87, 0) = 0$

- For $S = 199.35$: Payoff = $\max(182 - 199.35, 0) = 0$

- For $S = 180$: Payoff = $\max(182 - 180, 0) = 2$

- For $S = 162.53$: Payoff = $\max(182 - 162.53, 0) = 19.47$

- For $S = 146.70$: Payoff = $\max(182 - 146.70, 0) = 35.30$


**Now we do Backward Induction as our 6th step**

Now here we'll work backward to calculate the present value of the option at each step and thats discounting the future payoffs.

The formula for backward induction  givrn by--

\
$$
V = e^{-r \Delta t} \times (p \times V_{\text{up}} + (1 - p) \times V_{\text{down}})
$$

\
- **We start by Step 2 to Step 1** (for $S = 199.35$, $S = 180$, $S = 162.53$):

  - For $S = 199.35$:

    $$
    V_{199.35} = e^{-0.02 \times 0.1667} \times (0.4908 \times 0 + 0.5092 \times 0) = 0
    $$


  - For $S = 180$:

    $$
    V_{180} = e^{-0.02 \times 0.1667} \times (0.4908 \times 0 + 0.5092 \times 2) = 1.01
    $$

  \

  - For $S = 162.53$:

    $$
    V_{162.53} = e^{-0.02 \times 0.1667} \times (0.4908 \times 2 + 0.5092 \times 19.47) = 10.76
    $$

  \

- **And now the Step 1 to Step 0** (for $S_0 = 180$):
  - For $S_0 = 180$:
    $$
    V_0 = e^{-0.02 \times 0.1667} \times (0.4908 \times 1.01 + 0.5092 \times 10.76) = 5.96
    $$

#### As we see the price of the European put option using the 3-step binomial tree is **5.96**.


### Part B. Delta Hedging Process

**Quiz i. Pick One Path in the Tree**

We will pick the path where the stock goes Up → Down → Down.

**Quiz part 25 A ii. Describe the Delta Hedging Process**

Lets describe the process step by step

**First as Step 1:We take the Initial Delta at $t = 0$**

At time $t = 0$, Lets calculate the Delta using the stock price $S_0 = 180$, and the possible next steps $S_u(u=1.1075)$ and $S_d(d=0.9029)$.

**So the Stock Prices are**
- $S(u) = 199.35$
- $S(d) = 162.53$

** and the Payoffs are**
- $V(u) = 0$ (option is out-of-the-money)
- $V(d) = 10.76$ (As we calculated from backward induction)

**Now we do the Delta Calculation which is -**

\
$$
\Delta_0 = \frac{V(u) - V(d)}{S(u) - S(d)} = \frac{0 - 10.76}{199.35 - 162.53} = \frac{-10.76}{36.82} = -0.2923
$$

\

####This means that you sell 0.2923 shares of the underlying asset to hedge the option.


**Our Step 2: Lets see the After Stock Price Moves Up**

when the stock price moves up to $S(u) = 199.35$ so therefore we'll now calculate Delta again based on the new possible stock prices.

**Stock Prices**
- $S(u) = 220.87$
- $S(d) = 180$

**Payoffs are now-**
- $V(u) = 0$ (option is out-of-the-money)
- $V(d) = 1.01$

**Next we do the Delta Calculation again**

\
$$
\Delta_1 = \frac{V(u) - V(d)}{S(u) - S(d)} = \frac{0 - 1.01}{220.87 - 180} = \frac{-1.01}{40.87} = -0.0247
$$

\

Here at this step you'll sell 0.0247 more shares of the underlying asset.


**To our Step 3: Now lets see After the Stock Price Moves Down**

When the stock price moves down to $S(d) = 180$; we'll now calculate Delta again using the updated prices.

**The Stock Prices become**
- $S(u) = 180$
- $S(d) = 162.53$

** While the Payoffs become-**
- $V(u) = 1.01$
- $V(d) = 19.47$

**Again doing the Delta Calculation we get**

\
$$
\Delta_2 = \frac{V(u) - V(d)}{S(u) - S(d)} = \frac{1.01 - 19.47}{180 - 162.53} = \frac{-18.46}{17.47} = -1.0567
$$

\

And now at this step you will sell 1.0567 shares of the underlying asset.

###**25 A part iii. Table of Cash Account Evolution**

We will now show the evolution of the cash account based on the dynamic Delta hedging process at each step.

**Lets start with the Initial Cash Position at $t = 0$:**

If we sell $\Delta_0 = -0.2923$ shares of the underlying asset then the initial cash position becomes-

$$
\text{Cash at } t = 0 = 0.2923 \times 180 = 52.61
$$

**The table below will show the Evolution of Cash Account**

\[
\begin{array}{|c|c|c|c|c|}
\hline
\text{Step} & \text{Stock Price} & \text{Delta} & \text{Units Bought/Sold} & \text{Cash Account} \\
\hline
0 & 180 & -0.2923 & \text{We Sell 0.2923 shares} & 52.61 \\
\hline
1 & 199.35 & -0.0247 & \text{We Sell 0.0247 more shares} & 52.61 + (0.0247 \times 199.35) = 57.54 \\
\hline
2 & 180 & -1.0567 & \text{Also here we Sell 1.0567 more shares} & 57.54 + (1.0567 \times 180) = 247.74 \\
\hline
\end{array}
\]

We then deduct taht at each step we will sell more shares of the underlying asset and accumulate cash in your account.

The final cash account value then becomes= **247.74**.


##26.Using the same data from Q25, price an American Put option.

we need to:-

- Create a 25-step binomial tree for the stock prices.
- Price the American put option using backward induction, taking early exercise into account.
- Compute Delta at each node by calculating the change in option value for up and down movements.
-We will Use Python to automate the calculations.

And we know-


- $S_0 = 180$: Initial stock price
- $K = 182$: Strike price
- $r = 2\% = 0.02$: Risk-free interest rate
- $\sigma = 25\% = 0.25$: Volatility
- $T = 6$ months = 0.5 years: Time to maturity
- 25 steps in the binomial tree


###Lets's use the code below- Pricing the American Put Option and Delta Hedging

In [ ]:
import numpy as np

# Our Parameters as defined
S0 = 180  # Initial stock price
K = 182   # Strike price
r = 0.02  # Risk-free interest rate
T = 0.5   # Time to maturity (6 months)
sigma = 0.25  # Volatility
N = 25  # Number of steps in the tree

# Step size and probabilities
dt = T / N
u = np.exp(sigma * np.sqrt(dt))  # Up factor
d = 1 / u  # Down factor

# Display calculated u and d values
print(f"Up factor (u): {u:.4f}")
print(f"Down factor (d): {d:.4f}")

p = (np.exp(r * dt) - d) / (u - d)  # Risk-neutral probability

# Initialize stock price and option value arrays
stock_prices = np.zeros((N+1, N+1))
put_values = np.zeros((N+1, N+1))

# Calculate stock prices at each node
for i in range(N+1):
    for j in range(i+1):
        stock_prices[j, i] = S0 * (u**(i-j)) * (d**j)
        stock_prices[j, i] = round(stock_prices[j, i], 2)  # Round stock prices to nearest cent

# Calculate option payoffs at maturity
for j in range(N+1):
    put_values[j, N] = max(K - stock_prices[j, N], 0)
    put_values[j, N] = round(put_values[j, N], 2)  # Round option payoffs to nearest cent

# Backward induction for option price and early exercise condition
for i in range(N-1, -1, -1):
    for j in range(i+1):
        hold_value = np.exp(-r * dt) * (p * put_values[j, i+1] + (1-p) * put_values[j+1, i+1])
        put_values[j, i] = max(hold_value, K - stock_prices[j, i])  # Early exercise condition
        put_values[j, i] = round(put_values[j, i], 2)  # Round option prices to nearest cent

# American put option price at time 0
american_put_price = put_values[0, 0]
print(f"American Put Option Price: {american_put_price:.2f}")

# Delta calculation at each node
delta_hedge = np.zeros((N, N))
for i in range(N):
    for j in range(i+1):
        delta_hedge[j, i] = (put_values[j, i+1] - put_values[j+1, i+1]) / (stock_prices[j, i] - stock_prices[j+1, i])
        delta_hedge[j, i] = round(delta_hedge[j, i], 4)  # Round Delta to 4 decimal places

# Display Delta hedge values at the first few steps for clarity
print("Delta hedge values at the first 5 steps:")
print(delta_hedge[:5, :5])


Up factor (u): 1.0360
Down factor (d): 0.9653
American Put Option Price: 13.03
Delta hedge values at the first 5 steps:
[[-0.0337 -0.4085 -0.3268 -0.2504 -0.1823]
 [ 0.     -0.0397 -0.4963 -0.4085 -0.323 ]
 [ 0.      0.     -0.0458 -0.5885 -0.4996]
 [ 0.      0.      0.     -0.0518 -0.6812]
 [ 0.      0.      0.      0.     -0.0575]]


_**American Put Option Price:**_

From out code above-We have calculated the price of the American put option is $13.03$

This is to means that at time $t = 0$ the American put option will be worth 13.03.

**Now lets us do the Calculation for the Delta Hedging**

We know that Delta-measures how sensitive the price of the option is to changes in the price of the underlying asset.
As the seller of the option you"ll need adjust your position by buying or selling the underlying asset according to Delta at each step.

Here are the Delta hedge values at the first five steps-

\[
\begin{array}{|c|c|c|c|}
\hline
\text{Step} & \text{Path (Up/Down)} & \text{Stock Price} (S) & \text{Delta Hedge Value} (\Delta) \\
\hline
0 & \text{Initial} (S_0) & 180.00 & -0.0337 \\
1 & \text{Up($u$)} & 186.48 & -0.4085 \\
1 & \text{Down($d$)} & 173.74 & -0.0397 \\
2 & \text{Up-Up($u-u$)} & 193.18 & -0.3268 \\
2 & \text{Up-Down} & 180.00 & -0.4963 \\
2 & \text{Down-Down} & 167.57 & -0.0458 \\
3 & \text{Up-Up-Up} & 200.13 & -0.2504 \\
3 & \text{Up-Up-Down} & 186.48 & -0.5885 \\
3 & \text{Down-Down-Down} & 161.79 & -0.6812 \\
\hline
\end{array}
\]

**Lets Interprete theDelta Hedge Values:**

- At **Step 0** (thats is when theinitial stock price $S_0 = 180$)-Delta is -0.0337. This is to say that you will sell 0.0337 shares of the underlying asset to hedge the American put option.

-Next at **Step 1 (up)** (Thats is to say when the stock price increases to $S(u) = 199.35$) the delta is -0.4085 and you will sell an additional 0.4085 shares.
- On the other hand at **Step 1 (down)** (and this is to say when the stock price decreases to $S(d) = 162.53$) the delta becomes -0.0397 and this is to mean that You will sell an additional 0.0397 shares.

This process will continues at each step adjusting the hedge based on the updated stock prices and option values.

**Conclusion:**

- We have seen that the American put option price is **13.03** and this is reflecting the value of the option at time $t = 0$.

- The Delta hedging values will show how much of the underlying asset you need to buy or sell at each step to maintain a hedge.

The Delta  are  ranging from small adjustments in early steps (-0.0337 at Step 0) to larger adjustments as you move further down the tree.


**26 (b): Evolution of the Cash Account Throughout the Steps for One Path**

Here in this step we will track the cash-account evolution for a specific path in the binomial tree while performing Delta hedging.

The cash account will reflects the changes as we dynamically buy and sell shares of the underlying stock to maintain the hedge.

We will choose the path where the stock price moves **up → down → down** over the first three steps. This is to  means the stock price first increases, then decreases twice.

**SO our Path Selection is ** Up → Down → Down

**Steps:**
- **Step 0:** Initial price $S_0 = 180$
- **Step 1:** Stock  will moves up to 186.48
- **Step 2:** Stock price will moves down to 180
- **Step 3:** Stock price will then moves down again to 167.57

**Cash Account Evolution:**

The cash account will be influenced by:
1. The Delta at each step.
2. The stock price change ( this is which will dictates how much we buy or sell of the underlying stock.
3. The Interest earned on the cash account (We are discounting future values).

We will us The formula blw for updating the cash account-

$$
\text{Cash}_{t+1} = \text{Cash}_t \times e^{r \Delta t} + (\Delta_t - \Delta_{t+1}) \times S_{t+1}
$$

Where:
- $\text{Cash}_t$   -> is the cash position at time $t$.
- $\Delta_t$   ->is the Delta at time $t$, which tells us how many shares we hold at that step.
- $S_{t+1}$    ->is the stock price at time $t+1$.
- $r = 2\%$     ->is the risk-free rate.


In [ ]:
import numpy as np

# Parameters
S0 = 180  # Initial stock price
K = 182   # Strike price
r = 0.02  # Risk-free interest rate
T = 0.5   # Time to maturity (6 months)
sigma = 0.25  # Volatility
N = 25  # Number of steps in the tree

# Step size and probabilities
dt = T / N
u = np.exp(sigma * np.sqrt(dt))  # Up factor
d = 1 / u  # Down factor
p = (np.exp(r * dt) - d) / (u - d)  # Risk-neutral probability

# Initialize stock price and option value arrays
stock_prices = np.zeros((N+1, N+1))
put_values = np.zeros((N+1, N+1))

# Calculate stock prices at each node
for i in range(N+1):
    for j in range(i+1):
        stock_prices[j, i] = S0 * (u**(i-j)) * (d**j)
        stock_prices[j, i] = round(stock_prices[j, i], 2)  # Round stock prices to nearest cent

# Calculate option payoffs at maturity
for j in range(N+1):
    put_values[j, N] = max(K - stock_prices[j, N], 0)
    put_values[j, N] = round(put_values[j, N], 2)  # Round option payoffs to nearest cent

# Backward induction for option price and early exercise condition
for i in range(N-1, -1, -1):
    for j in range(i+1):
        hold_value = np.exp(-r * dt) * (p * put_values[j, i+1] + (1-p) * put_values[j+1, i+1])
        put_values[j, i] = max(hold_value, K - stock_prices[j, i])  # Early exercise condition
        put_values[j, i] = round(put_values[j, i], 2)  # Round option prices to nearest cent

# American put option price at time 0
american_put_price = put_values[0, 0]
print(f"American Put Option Price: {american_put_price:.2f}")

# Delta calculation at each node
delta_hedge = np.zeros((N, N))
for i in range(N):
    for j in range(i+1):
        delta_hedge[j, i] = (put_values[j, i+1] - put_values[j+1, i+1]) / (stock_prices[j, i] - stock_prices[j+1, i])
        delta_hedge[j, i] = round(delta_hedge[j, i], 4)  # Round Delta to 4 decimal places

# Path chosen: Up -> Down -> Down
path_stock_prices = [S0, stock_prices[0, 1], stock_prices[1, 2], stock_prices[2, 3]]
path_delta = [delta_hedge[0, 0], delta_hedge[0, 1], delta_hedge[1, 2], delta_hedge[2, 3]]

# Initialize cash account
cash_account = np.zeros(4)
cash_account[0] = path_delta[0] * S0  # Initial cash position

# Calculate cash account evolution
for t in range(1, 4):
    cash_account[t] = cash_account[t-1] * np.exp(r * dt) + (path_delta[t-1] - path_delta[t]) * path_stock_prices[t]

# Display results
print("\nCash Account Evolution for Path: Up -> Down -> Down\n")
for t in range(4):
    print(f"Step {t}: Stock Price = {path_stock_prices[t]:.2f}, Delta = {path_delta[t]:.4f}, Cash Account = {cash_account[t]:.2f}")


American Put Option Price: 13.03

Cash Account Evolution for Path: Up -> Down -> Down

Step 0: Stock Price = 180.00, Delta = -0.0337, Cash Account = -6.07
Step 1: Stock Price = 186.48, Delta = -0.4085, Cash Account = 63.82
Step 2: Stock Price = 180.00, Delta = -0.4963, Cash Account = 79.65
Step 3: Stock Price = 173.75, Delta = -0.5885, Cash Account = 95.71


**our path- Up → down → down at each step:**

\[
\begin{array}{|c|c|c|c|}
\hline
\text{Step} & \text{Stock Price (S)} & \text{Delta (Δ)} & \text{Cash Account (Cash)} \\
\hline
0 & 180.00 & -0.0337 & -6.07 \\
1 & 186.48 & -0.4085 & 63.82 \\
2 & 180.00 & -0.4963 & 79.65 \\
3 & 173.75 & -0.5885 & 95.71 \\
\hline
\end{array}
\]


**At step 0 (the Initial):**

- We see thatthe stock price starts at $S_0 = 180.00$ while the Delta is $-0.0337$.  
This is to means that you need to sell 0.0337 shares of the stock to hedge the option. And since the Delta is negative, you will sell shares and the initial cash account balance will become $-6.07$.
The negative balance reflect the fact that you started by selling a small portion of shares.

**Now at Step 1 (Up($u$)):**

- We have the stock price increasing to 186.48 while the Delta changes to $-0.4085$.  
Here you will sell more shares (the Delta is more negative) t which will help us to maintain a neutral hedge. After selling more shares the cash account will increases to $63.82$. This rise in the cash balance is  comeing from selling additional shares at a higher price.

**At Step 2 (Down Movement):**

-Here the stock price returns to 180.00 while and the Delta becomes $-0.4963$.  
The stock price has decreased and this is to mean that you'll need to sell slightly more shares to adjust for the change in Delta.
As a result the cash account also further increases to $79.65$. The increase in cash is due to selling more shares even though the stock price has returned to its initial level.

**At Step 3 (This is the second Down Movement):**

-Also here the stock price will falls to 173.75 while the Delta now changes to $-0.5885$.  
The stock price decreases and thus this means that you will sell even more shares to hedge against further price drops and this is bringing the cash account up to $95.71$. By selling shares as the stock price declines the hedge will continues to build up cash reserves.


The cash account is a reflection of  the proceeds from continuously adjusting the hedging position by selling shares of the underlying stock as the Delta changes at each step. In a Delta-hedged portfolio our objective is to offset changes in the value of the option with changes in the value of the underlying stock. And now here as the stock price fluctuates the Delta will becomes more negative and this means taht more shares are sold to maintain a neutral position.

By the end of the third step the cash account will have  accumulated $95.71$. This positive balance is demonstrating how selling shares to adjust the hedge has resulted in accumulating cash over time(even as the stock price moves along the chosen path).



 **26 (c): Comment on the Delta Hedging Process as Compared to the European Option Case**

###**Some of the Key Observations from Our Work:**

**Delta Adjustments for American Put Option:**

We observe/see that the Delta for the American put option will fluctuate more at each step.As the stock price followed the path up → down → down the Delta moved from -0.0337 at Step 0 to -0.5885.

There are sharper changes in Delta. This reflects an early exercise inherent in American options more so as the stock price approached the strike price and the option moves closer to being in-the-money (Hull, 2018).

Early exercise will introduce complexity into the hedging process which in turn requires larger adjustments to maintain the Delta-neutral position.

As a result the rapid changes in Delta will require more frequent and substantial adjustments in the number of shares held which inturns leads to larger swings in the cash account and in our case which increased from -6.07 to 95.71 over three steps.

**Delta Adjustments for European Put Option:**

The Delta for the European put option (which we priced in Step 25 above) evolved more smoothly. This is because European options are only be exercised at maturity which will eliminates the risk of early exercise (Wilmott, 2006).

Without the early exercise the Delta for the European put option will follow a more gradual path as the stock price moves which will be requiring fewer sharp adjustments in the hedging position.


